rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,296 @@
+# RFC 120 object Service
+* Comments: [#116](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/120)
+* Proposed by: @jhford
+
+# Summary
+
+Taskcluster tasks often produce artifacts during the course of being run.
+Examples include browser distribution archives, json metadata and log files.
+We currently store artifacts in a single region of S3 and split management of
+artifacts between the Queue and Cloud Mirror.  This project is to design an
+object storage api which drastically simplifies the management of objects from
+the Queue's perspective while at the same time handling the cross-region
+replication and cache management of Cloud Mirror as well as enabling
+cross-cloud support.
+
+The object service should not be directly exposed to end users as a Rest API.
+It will instead be a generalised API that the Queue uses to facilitate artifact
+creation from tasks.  In other words, the Queue gates operations to the object
+service.  The service will be exposed to endusers by means of a `Location`
+header value on a `300` series redirect or by relaying responses.
+
+The Queue will now store only the name of objects and would treat the name of
+an object as an opaque identifier and defer all management of the actual bytes
+to this system.  The Queue would also stop doing IP to object store mappings.
+
+This RFC is to decide on what the objects service is and its interface, not
+implementation details.
+
+## Motivation
+
+* We want the Queue to be easier to understand, develop and maintain
+* We want objects to be easier to understand, develop and maintain
+* We want better object handling without increasing complexity of the Queue
+  codebase
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of objects into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
+# Details
+
+* Auth will not be done with scopes and tc-auth, but a shared secret between
+  frontend service and objects, with JWT being a possibility
+* The frontend service will block new artifacts after task resolution by not issuing new
+  signed urls to this service
```

Comment:
Which front end service? Should this be "block new objects" rather than "block new artifacts"? We should probably explain what generates signed urls and how they are passed between components, and then go on to say under which circumstances, a request for a signed URL will be denied (e.g. by returning an HTTP 403 status code to a given HTTP request).

Reply:
Nope, it's the Queue which would be doing the blocking here.  So in this specific case, the blocking would be the queue checking the status of the task and then deciding to return a 400 series error response instead of a 200 with information on how to upload the object.